File: go/290.go
package main

import (
	"errors"
	"fmt"
	"strings"
)

func main() {
	if wordPattern("abba", "dog cat cat dog") != true ||
		wordPattern("abba", "dog dog dog dog") != false ||
		wordPattern("aaa", "aa aa aa aa") != false {
		panic(errors.New("Test case failed"))
	} else {
		fmt.Println("Done")
	}
}

func wordPattern(pattern string, s string) bool {
	runes := []rune(pattern)
	words := strings.Fields(s)

	if len(runes) != len(words) {
		return false
	}

	book1 := make(map[rune]string)
	book2 := make(map[string]rune)

	for i := 0; i < len(runes); i++ {
		r := runes[i]
		w := words[i]
		if book1[r] == "" {
			book1[r] = w
		}
		if book2[w] == 0 {
			book2[w] = r
		}

		if book1[r] != w || book2[w] != r {
			return false
		}
	}

	return true
}

File: python/290.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        s = s.split()
        if len(pattern) != len(s):
            return False

        book1 = {}
        book2 = {}

        for p, w in zip(pattern, s):
            if p not in book1:
                book1[p] = w
            if w not in book2:
                book2[w] = p

            if book1[p] != w or book2[w] != p:
                return False

        return True
